### PR TITLE
 opt: refactor limit hint propagation into a required physical property

### PR DESCRIPTION
**Pull request by savoie**
This patch follows up on #42086 and finishes the refactor of the
heuristic planner `applyLimits` function. It removes the last of the
heuristic planner limit handling, and introduces the LimitHint required
physical property, which is represented by a float64 in anticipation of
future improvements that will use statistics to inform better
estimations.

Previously, in the case where a limit hint was propagated down to a
limitNode that had an offset but no limit, the limit hint would be
discarded. In this patch, an OffsetOp's limit hint will be propagated to
its children, having been increased enough to ensure that the required
number of rows can be discarded by the offset.

Another change from the behaviour of the heuristic planner's handling of
limit hints is in the case of subqueries. `applyLimit` would not be
called on subqueries created with `Builder.addSubquery`, so no nodes
within subqueries had soft limits set. Since handling of required
physical properties occurs before subqueries are created, this ceases to
be a special case and soft limits may be set for scanNodes within
subqueries.

Besides these two departures from the heuristic planner's limit
propagation, this patch should not have introduced further changes
affecting the ultimate value of `scanNode.softLimit`. With this
refactor, improvements to limit hint handling can be made within the
optimizer.

Release note: None

### CODE REVIEW

**Comment by savoie** [comment]
If anyone has pro tips on how to make this easier to review with #42170 still unmerged, let me know :)

**Review by RaduBerinde — COMMENTED** [review]
Nice work! I posted various comments. Review was easy (using the "review each comment separately" setting in reviewable).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42170#-:-LstGS6RC4YrgNEZNBfh:b-wr2p6x)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @savoie)

___
*[pkg/sql/opt/exec/execbuilder/relational.go, line 471 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/42170#-LstA2Xc-yypeIez9TMq:-LstA2Xc-yypeIez9TMr:bsiy4ea) ([raw file](https://github.com/cockroachdb/cockroach/blob/9558ed8c6e9e287dbc2a1db80777c85fd3ee669c/pkg/sql/opt/exec/execbuilder/relational.go#L471)):*
> ```Go
> 	}
> 
> 	softLimit := int64(scan.RequiredPhysical().LimitHint)
> ```

Here we convert 0.1 to 0 (which is no limit). I think we should make it 1 if there's a limit hint.
___
*[pkg/sql/opt/props/physical/required.go, line 47 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/42170#-LstA9UOB36rpOvAb7H6:-LstA9UOB36rpOvAb7H7:b-cplojl) ([raw file](https://github.com/cockroachdb/cockroach/blob/9558ed8c6e9e287dbc2a1db80777c85fd3ee669c/pkg/sql/opt/props/physical/required.go#L47)):*
> ```Go
> 	Ordering OrderingChoice
> 
> 	LimitHint float64
> ```

Definitely needs a comment, especially mentioning the special value 0.
___
*[pkg/sql/opt/props/physical/required.go, line 86 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/42170#-LstBaj57mO43K5hPsgi:-LstBaj57mO43K5hPsgj:bgafctq) ([raw file](https://github.com/cockroachdb/cockroach/blob/9558ed8c6e9e287dbc2a1db80777c85fd3ee669c/pkg/sql/opt/props/physical/required.go#L86)):*
> ```Go
> 		buf.WriteByte(']')
> 
> 		if hasOrdering || hasLimitHint {
> ```

The logic in this function is very hard to follow (and hard to extend correctly). We should do something better, e.g.:

```go
var buf bytes.Buffer
output := func(name string, fn func (*bytes.Buffer)) {
  if buf.Len() != 0 {
    buf.WriteByte(' ')
  }
  buf.WriteByte('[')
  buf.WriteString(name)
  fn(&buf)
  buf.WriteByte(']')
}
```
and call `output` for each set property.
___
*[pkg/sql/opt/xform/optimizer.go, line 557 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/42170#-LstCQlQ7tga0yDciHdb:-LstCQlQ7tga0yDciHdc:bmfi08h) ([raw file](https://github.com/cockroachdb/cockroach/blob/9558ed8c6e9e287dbc2a1db80777c85fd3ee669c/pkg/sql/opt/xform/optimizer.go#L557)):*
> ```Go
> 	// enforcer operator can provide them.
> 	inner.Presentation = nil
> 	inner.LimitHint = 0
> ```

I don't think that conceptually the limit hint is the same as the presentation here. Zero just happens to be the right choice when adding a `SortExpr{}` below. And it's not even necessarily the best choice: we can pass through a limit hint in the second case below where we have a partial ordering. For example, if we are ordering by `a,b` and we have an ordering on `a` and we know that on average there are say 100 `b`s for every `a`, we can pass through `limitHint + 100`. This could be an important optimization (favoring using an index on `a` for this operation).

I think the right thing to do here is to calculate `inner` in the two cases below by calling `BuildChildPhysicalProps` on the `SortExpr` that we create. Then we will use whatever logic is in there (which we can improve later as suggested). It is already the case that `inner` here *must* be exactly what `BuildChildPhysicalProps` would return for the enforcer, it was just trivial to achieve that directly until now.
___
*[pkg/sql/opt/xform/physical_props.go, line 81 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/42170#-LstHV2684AS1fHbsBpD:-LstHV2684AS1fHbsBpE:b-53tijb) ([raw file](https://github.com/cockroachdb/cockroach/blob/9558ed8c6e9e287dbc2a1db80777c85fd3ee669c/pkg/sql/opt/xform/physical_props.go#L81)):*
> ```Go
> 	childProps.Ordering = ordering.BuildChildRequired(parent, &parentProps.Ordering, nth)
> 
> 	switch parent.Op() {
> ```

We need targeted tests for this logic in a `xform/testdata/physprops/limit-hint` file.
___
*[pkg/sql/opt/xform/physical_props.go, line 92 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/42170#-LstFTZh-OcSAGc_D73B:-LstFTZh-OcSAGc_D73C:b-tcqc62) ([raw file](https://github.com/cockroachdb/cockroach/blob/9558ed8c6e9e287dbc2a1db80777c85fd3ee669c/pkg/sql/opt/xform/physical_props.go#L92)):*
> ```Go
> 		if constOffset, ok := parent.(*memo.OffsetExpr).Offset.(*memo.ConstExpr); ok {
> 			offsetValue := float64(*constOffset.Value.(*tree.DInt))
> 			if offsetValue > math.MaxFloat64-parentProps.LimitHint {
> ```

Seems odd to have to worry about this for floats, I'd remove it. If you're really worried about it, you can add a catch-all below `if LimitHint == math.Inf(+1) { LimitHint = 0 }`
___
*[pkg/sql/opt/xform/physical_props.go, line 103 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/42170#-LstGGVg9xB-pdsU_t0w:-LstGGVg9xB-pdsU_t0x:b-7ryk0p) ([raw file](https://github.com/cockroachdb/cockroach/blob/9558ed8c6e9e287dbc2a1db80777c85fd3ee669c/pkg/sql/opt/xform/physical_props.go#L103)):*
> ```Go
> 	case opt.ExceptOp, opt.ExceptAllOp, opt.IntersectOp, opt.IntersectAllOp,
> 		opt.UnionOp, opt.UnionAllOp:
> 		childProps.LimitHint = parentProps.LimitHint
> ```

Add a TODO here to differentiate between the left and right child.
___
*[pkg/sql/opt/xform/physical_props.go, line 107 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/42170#-LstFzeF9PYEskbzWxL-:-LstFzeF9PYEskbzWxL0:boo7xq2) ([raw file](https://github.com/cockroachdb/cockroach/blob/9558ed8c6e9e287dbc2a1db80777c85fd3ee669c/pkg/sql/opt/xform/physical_props.go#L107)):*
> ```Go
> 		childProps.LimitHint = parentProps.LimitHint
> 	case opt.SelectOp:
> 		childProps.LimitHint = parentProps.LimitHint
> ```

Add a `TODO()` here that we should take into account the selectivity.
___
*[pkg/sql/opt/xform/physical_props.go, line 108 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/42170#-LstGLIO7lPSQmQ6u31k:-LstGLIO7lPSQmQ6u31l:bok60i9) ([raw file](https://github.com/cockroachdb/cockroach/blob/9558ed8c6e9e287dbc2a1db80777c85fd3ee669c/pkg/sql/opt/xform/physical_props.go#L108)):*
> ```Go
> 	case opt.SelectOp:
> 		childProps.LimitHint = parentProps.LimitHint
> 	case opt.ProjectOp:
> ```

these cases can be coalesced `case opt.ProjectOp, opt.OrdinalityOp, ...`

**Comment by savoie** [comment]
(Just adding a file I'd forgotten to bundle into the commit)

**Review by justinj — COMMENTED** [review]
Nice! Less code than I expected, but definitely shows that this was a subtle thing to work on.

It strikes me that it seems like conceptually what we really are tracking as the soft limit is really kind of a probability distribution of the number of rows we might need, and I wonder if future work would also track some other properties of this distribution (say, the soft limit itself is the mean, but then if we also track the stdev we can control how risk-averse we're being in a given situation).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42170#-:-LswVA4nCBO-YuxvlApT:b-z2umr8)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RaduBerinde and @savoie)

___
*[pkg/sql/opt/xform/physical_props.go, line 107 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/42170#-LstFzeF9PYEskbzWxL-:-LswT7BRDuLOqIfr8KC2:blmtua2) ([raw file](https://github.com/cockroachdb/cockroach/blob/9558ed8c6e9e287dbc2a1db80777c85fd3ee669c/pkg/sql/opt/xform/physical_props.go#L107)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Add a `TODO()` here that we should take into account the selectivity.
</blockquote></details>

Same with `DistinctOn`
___
*[pkg/sql/opt/xform/physical_props.go, line 113 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/42170#-LswTM5O2Pde45-FXTu5:-LswTM5O2Pde45-FXTu6:b-j6gpfa) ([raw file](https://github.com/cockroachdb/cockroach/blob/1eb99fa60bf87b5b452e14dd03eaf5fa408fefd3/pkg/sql/opt/xform/physical_props.go#L113)):*
> ```Go
> 		childProps.LimitHint = parentProps.LimitHint
> 	case opt.ProjectSetOp:
> 		childProps.LimitHint = parentProps.LimitHint
> ```

consider panicking here so we know to update it for new operators (even if we just have a catch-all case that contains all the existing operators)

**Review by savoie — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42170)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @justinj)

___
*[pkg/sql/opt/exec/execbuilder/relational.go, line 471 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/42170#-LstA2Xc-yypeIez9TMq:-Lt6s9ty4oVjEbyFOCFb:b-xbf2cu) ([raw file](https://github.com/cockroachdb/cockroach/blob/9558ed8c6e9e287dbc2a1db80777c85fd3ee669c/pkg/sql/opt/exec/execbuilder/relational.go#L471)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Here we convert 0.1 to 0 (which is no limit). I think we should make it 1 if there's a limit hint.
</blockquote></details>

Fixed by calling `math.Ceil` the 2 times we make the conversion to `int64`. If `LimitHint` needs to be converted to an integer number of rows many more times, it should probably get some dedicated `toNumberOfRows` function. For the moment though this seems fine.
___
*[pkg/sql/opt/xform/physical_props.go, line 81 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/42170#-LstHV2684AS1fHbsBpD:-Lt6se1iCAI9zVznPT-1:bdtpjpa) ([raw file](https://github.com/cockroachdb/cockroach/blob/9558ed8c6e9e287dbc2a1db80777c85fd3ee669c/pkg/sql/opt/xform/physical_props.go#L81)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

We need targeted tests for this logic in a `xform/testdata/physprops/limit-hint` file.
</blockquote></details>

Only have the one test right now (for the case of an `Offset` somewhere below a `Limit`), but will be adding more as we get into more complex hint calculations.
___
*[pkg/sql/opt/xform/physical_props.go, line 113 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/42170#-LswTM5O2Pde45-FXTu5:-Lt6sw-e6xgLQP1x_y2U:bhubt9n) ([raw file](https://github.com/cockroachdb/cockroach/blob/1eb99fa60bf87b5b452e14dd03eaf5fa408fefd3/pkg/sql/opt/xform/physical_props.go#L113)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

consider panicking here so we know to update it for new operators (even if we just have a catch-all case that contains all the existing operators)
</blockquote></details>

Radu and I discussed this and it seems not worth it since the default case here ("don't impose a limit hint on your children") is maybe suboptimal in some cases but never _wrong_.

**Review by rytaft — COMMENTED** [review]
Nice work!

> Reviewed 12 of 16 files at r2, 1 of 1 files at r3, 5 of 5 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42170#-:-Lt7dyDG6jrXoWZfg0-e:bdtx8r6)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @justinj and @savoie)

___
*[pkg/sql/opt/exec/execbuilder/relational.go, line 475 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/42170#-Lt7_SE_DWHOlcVc_tHP:-Lt7_SE_DWHOlcVc_tHQ:bw6vdk) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a3f09a54d32f11b683c6971a40dcbcc49660c74/pkg/sql/opt/exec/execbuilder/relational.go#L475)):*
> ```Go
> 	hardLimit := scan.HardLimit.RowCount()
> 
> 	// At most one of hardLimit and softLimit may be defined at the same time.
> ```

why is softLimit prioritized over hardLimit?
___
*[pkg/sql/opt/memo/interner.go, line 539 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/42170#-Lt7b6nvDWp0gdY7Bm9M:-Lt7b6nvDWp0gdY7Bm9N:b-mvoc6d) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a3f09a54d32f11b683c6971a40dcbcc49660c74/pkg/sql/opt/memo/interner.go#L539)):*
> ```Go
> 	}
> 	h.HashOrderingChoice(val.Ordering)
> 	h.HashFloat64(val.LimitHint)
> ```

Is there any chance we could have a hash collision here with a longer ordering choice but no limit hint? 
___
*[pkg/sql/opt/props/physical/required.go, line 52 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/42170#-Lt7bcxLAOpaA6YgB6Bk:-Lt7bcxLAOpaA6YgB6Bl:bnpdluq) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a3f09a54d32f11b683c6971a40dcbcc49660c74/pkg/sql/opt/props/physical/required.go#L52)):*
> ```Go
> 	// to return all result rows, but it can be optimized based on the assumption
> 	// that only the hinted number of rows will be needed.
> 	// A LimitHint of 0 indicates "no limit". The LimitHint an intermediate
> ```

[nit] LimitHint has an intermediate

**Review by savoie — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42170)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @justinj and @rytaft)

___
*[pkg/sql/opt/exec/execbuilder/relational.go, line 475 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/42170#-Lt7_SE_DWHOlcVc_tHP:-LtBDvot33J4bob-ImAP:b1j5vna) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a3f09a54d32f11b683c6971a40dcbcc49660c74/pkg/sql/opt/exec/execbuilder/relational.go#L475)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

why is softLimit prioritized over hardLimit?
</blockquote></details>

Overwriting a hard limit with a soft limit was the default behaviour in `applyLimits` and I was aiming for as close to a straight refactor as possible with this patch. It's not immediately obvious to me what is optimal if, say, a node has a large hard limit and a small soft limit. Adding a TODO here.
___
*[pkg/sql/opt/memo/interner.go, line 539 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/42170#-Lt7b6nvDWp0gdY7Bm9M:-LtBe089BXg6U5FkqNP9:b-4qn84t) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a3f09a54d32f11b683c6971a40dcbcc49660c74/pkg/sql/opt/memo/interner.go#L539)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Is there any chance we could have a hash collision here with a longer ordering choice but no limit hint? 
</blockquote></details>

OK so this was an interesting can of worms. I think the answer is probably yes, we could have a collision, but before I got to constructing one, I found out that it's actually quite simple to construct a hash collision without even adding any properties besides `OrderingChoice`: `-2 opt(1)` and `-(1|2)` have the same hash. This is handled correctly by `InternPhysicalProps`, which does verify that physical properties with the same hash are actually equal before considering them duplicates.

So I think the entirety of `HashPhysProps` would need a rewrite to avoid collisions, but also collisions are handled correctly, so infrequent collisions look like they're probably fine?

(Side note: turns out we weren't running all of the physical props interner tests; fixing that in this new push.)
___
*[pkg/sql/opt/props/physical/required.go, line 52 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/42170#-Lt7bcxLAOpaA6YgB6Bk:-LtBh2enEwpaatSqIgdt:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a3f09a54d32f11b683c6971a40dcbcc49660c74/pkg/sql/opt/props/physical/required.go#L52)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

[nit] LimitHint has an intermediate
</blockquote></details>

Done.

**Review by andy-kimball — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42170)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @justinj and @rytaft)

___
*[pkg/sql/opt/memo/interner.go, line 539 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/42170#-Lt7b6nvDWp0gdY7Bm9M:-LtC4mc_6BVjYSYgSrem:buaz82f) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a3f09a54d32f11b683c6971a40dcbcc49660c74/pkg/sql/opt/memo/interner.go#L539)):*
<details><summary><i>Previously, savoie (Céline O'Neil) wrote…</i></summary><blockquote>

OK so this was an interesting can of worms. I think the answer is probably yes, we could have a collision, but before I got to constructing one, I found out that it's actually quite simple to construct a hash collision without even adding any properties besides `OrderingChoice`: `-2 opt(1)` and `-(1|2)` have the same hash. This is handled correctly by `InternPhysicalProps`, which does verify that physical properties with the same hash are actually equal before considering them duplicates.

So I think the entirety of `HashPhysProps` would need a rewrite to avoid collisions, but also collisions are handled correctly, so infrequent collisions look like they're probably fine?

(Side note: turns out we weren't running all of the physical props interner tests; fixing that in this new push.)
</blockquote></details>

I tried to keep the hashing functions extremely simple and fast. Collisions aren't a problem unless they're common.

**Comment by RaduBerinde** [comment]
___
*[pkg/sql/opt/memo/interner.go, line 539 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/42170#-Lt7b6nvDWp0gdY7Bm9M:-LtCYcaM8h1pB1xA_zTo:b56gjb9) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a3f09a54d32f11b683c6971a40dcbcc49660c74/pkg/sql/opt/memo/interner.go#L539)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I tried to keep the hashing functions extremely simple and fast. Collisions aren't a problem unless they're common.
</blockquote></details>

Yes, our approach is to keep things simple, but avoid unnecessary collisions if it's not more complicated. To answer thr original concern, because we "append" the float in all cases, it shouldn't be a problem. It would be if we only hashed it when it's nonzero.

**Review by rytaft — APPROVED** [review]
:lgtm:

> Reviewed 3 of 3 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42170#-:-LtRZpjl3eKEYRzEOctq:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @justinj)

___
*[pkg/sql/opt/memo/interner.go, line 539 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/42170#-Lt7b6nvDWp0gdY7Bm9M:-LtRZiVnAEa8MOd9PuQD:b-xe5y6l) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a3f09a54d32f11b683c6971a40dcbcc49660c74/pkg/sql/opt/memo/interner.go#L539)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Yes, our approach is to keep things simple, but avoid unnecessary collisions if it's not more complicated. To answer thr original concern, because we "append" the float in all cases, it shouldn't be a problem. It would be if we only hashed it when it's nonzero.
</blockquote></details>

Makes sense - thanks!

**Review by RaduBerinde — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42170#-:-Lt__-rB3ugxtN8fPx1l:bnfp4nl)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @justinj and @savoie)

___
*[pkg/sql/opt/xform/testdata/physprops/limit_hint, line 4 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/42170#-Lt_ZkLLD9eu59QmJt9b:-Lt_ZkLMBq7dvqDgNW-E:b-kqp59p) ([raw file](https://github.com/cockroachdb/cockroach/blob/800913599f2b3faf3a6f31bb65f7a3c0919b5f56/pkg/sql/opt/xform/testdata/physprops/limit_hint#L4)):*
> ```
> CREATE TABLE t (x INT PRIMARY KEY, y INT)
> ----
> ```

I would add some simple tests with `opt` that show that the hint gets all the way into the scan, including through index join and other cases where we're passing the limit down.

**Review by savoie — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42170)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @justinj and @rytaft)

___
*[pkg/sql/opt/xform/testdata/physprops/limit_hint, line 4 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/42170#-Lt_ZkLLD9eu59QmJt9b:-LtgBsUE6SybxFu78Rb8:b-3bwk6v) ([raw file](https://github.com/cockroachdb/cockroach/blob/800913599f2b3faf3a6f31bb65f7a3c0919b5f56/pkg/sql/opt/xform/testdata/physprops/limit_hint#L4)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I would add some simple tests with `opt` that show that the hint gets all the way into the scan, including through index join and other cases where we're passing the limit down.
</blockquote></details>

Added the limit hint to the output of `opt` and added some simple tests that cover this.

**Review by justinj — COMMENTED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42170#-:-LtQmFJr4HhPKOGslowl:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 2 stale) (waiting on @justinj and @rytaft)

___
*[pkg/sql/opt/exec/execbuilder/relational.go, line 475 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/42170#-Lt7_SE_DWHOlcVc_tHP:-LtQluU04V0UG-aYN1TN:bh7yvr6) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a3f09a54d32f11b683c6971a40dcbcc49660c74/pkg/sql/opt/exec/execbuilder/relational.go#L475)):*
<details><summary><i>Previously, savoie (Céline O'Neil) wrote…</i></summary><blockquote>

Overwriting a hard limit with a soft limit was the default behaviour in `applyLimits` and I was aiming for as close to a straight refactor as possible with this patch. It's not immediately obvious to me what is optimal if, say, a node has a large hard limit and a small soft limit. Adding a TODO here.
</blockquote></details>

if we have a `hardLimit` won't we almost always have a `softLimit`? Since the softLimit will propagate down from a `Limit` operator?

**Review by savoie — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42170)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 2 stale) (waiting on @justinj and @rytaft)

___
*[pkg/sql/opt/exec/execbuilder/relational.go, line 475 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/42170#-Lt7_SE_DWHOlcVc_tHP:-LtzDpNl6DBBaNnEC1YJ:b-okjhb5) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a3f09a54d32f11b683c6971a40dcbcc49660c74/pkg/sql/opt/exec/execbuilder/relational.go#L475)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

if we have a `hardLimit` won't we almost always have a `softLimit`? Since the softLimit will propagate down from a `Limit` operator?
</blockquote></details>

`hardLimit` gets added in exploration rules, which replace a `Limit` around a `Scan` with a hard-limited `Scan`, so in that case there is no longer a `Limit` operator that can propagate down a limit hint, and there is only a hard limit.

The case of competing hard and soft limits only comes up if there were at least 2 limits above the scan, one of which can't be pushed down into a scan:

```
opt
SELECT * FROM (SELECT * FROM t LIMIT 20) where y=1 LIMIT 10
----
limit
 ├── columns: x:1(int!null) y:2(int!null) z:3(int)
 ├── select
 │    ├── columns: x:1(int!null) y:2(int!null) z:3(int)
 │    ├── limit hint: 10.00
 │    ├── scan t
 │    │    ├── columns: x:1(int!null) y:2(int) z:3(int)
 │    │    ├── limit: 20
 │    │    └── limit hint: 10.00
 │    └── filters
 │         └── y = 1 [type=bool]
 └── const: 10 [type=int]
```

If the hint of 10 were actually any good, it might be better to take that over the hard limit of 20. Once I have better limit hint estimates in cases like filtering, I'll look into whether something like "take the lowest of the hard and soft limit" is good in most cases.

**Review by RaduBerinde — APPROVED** [review]
:lgtm:

> Reviewed 7 of 16 files at r2, 3 of 5 files at r4, 2 of 3 files at r5, 47 of 47 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42170#-:-LtzWLWq2pmLiYZbVKhA:bnfp4nl)* status: :shipit: complete! 2 of 0 LGTMs obtained (and 1 stale) (waiting on @justinj)
